src/storage/src/render/upsert.rs

```diff
@@ -491,16 +491,30 @@ fn process_new_data(
         value: new_value,
         position: new_position,
         upstream_time_millis: _,
-        partition: _,
+        partition,
         metadata,
     } in new_data.drain(..)
     {
         let mut time = cap.time().clone();
         time.advance_by(as_of_frontier.borrow());
-        if key.is_none() {
-            error!(?new_value, "Encountered empty key for value");
-            continue;
-        }
+
+        let key = match key {
+            Some(key) => Some(key),
+            None => {
+                // NOTE: We need to make sure that when we see a "retraction" for this, that is, a
+                // message with _both_ a NULL key and NULL value, we can match this to the original
+                // `DecodeError`. Therefore, we cannot put more information in here than the
+                // partition. It would seem desirable to have more information, for example the
+                // offiset, but with that the future "retraction" message will not match this
+                // message and the original error will not go away.
```

Comment:
we really need to structure these errors someday!
One nice thing is if we could forward on some extra context with the error, and use that in the `details` column in `mz_source_status_history`...might be a good project for a new member!

Reply:
A hotly debated topic, but currently we don't show decoding errors in the source status tables at all. (Since the source can make progress even if some records fail to decode.) Seems worth it to capture the extra context regardless of how that shakes out though!

Reply:
I don't think these errors should be in the status history, they are an error that is deterministically derivable/derived from the data, so should live in the data collections